demo/llama/strategies.py

```diff
@@ -0,0 +1,241 @@
+            self.checkpoint_group = dist.new_group(
+                default_ranks, backend=self.process_group_backend)
+
+    def save_checkpoint(
+        self,
+        path: _PATH,
+        state: Dict[str, Union[Module, Optimizer, Any]],
+        storage_options: Optional[Any] = None,
+        filter: Optional[Dict[str, Callable[[str, Any], bool]]] = None,
+    ) -> None:
+        if storage_options is not None:
+            raise TypeError(
+                "`FSDPStrategy.save_checkpoint(..., storage_options=...)` is\
+                not supported because`FSDPStrategy` does not use the \
+                    `CheckpointIO`.")
+        from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+        modules = [
+            module for module in state.values() if _has_fsdp_modules(module)
+        ]
+        if len(modules) == 0:
+            raise ValueError(
+                "Could not find a FSDP model in the provided checkpoint state. Please provide the model as"
+                " part of the state like so: `save_checkpoint(..., state={'model': model, ...})`. Make sure"
+                " you set up the model (and optimizers if any) through the strategy before saving the checkpoint."
+            )
+        if len(modules) > 1:
+            raise ValueError(
+                "Found multiple FSDP models in the given state. Saving checkpoints with FSDP is"
+                " currently limited to a single model per checkpoint. To save multiple models, call the"
+                " save method for each model separately with a different path."
+            )
+        module = modules[0]
+        state_dict_ctx = _get_sharded_state_dict_context(module)
+
+        converted_state: Dict[str, Any] = {}
+        metadata: Dict[str, Any] = {}
+        with state_dict_ctx:
+            for key, obj in state.items():
+                converted: Any
+                if isinstance(obj, Module):
+                    converted = obj.state_dict()
+                    target_dict = converted_state
+                elif isinstance(obj, Optimizer):
+                    converted = FSDP.optim_state_dict(module, obj)
+                    target_dict = converted_state
+                else:  # everything not a module or optimizer is considered metadata
+                    converted = obj.state_dict() if isinstance(
+                        obj, _Stateful) else obj
+                    target_dict = metadata
+                _apply_filter(key, filter or {}, converted, target_dict)
+
+        path = Path(self.broadcast(path))
+        writer = GCSDistributedWriter(path, self.storage_client.project,
+                                      self.storage_client)
+
+        start_time = time.time()
+        if self.use_async:
+            self._async_save(converted_state, path, writer)
+        else:
+            self._save(converted_state, path, writer)
+        duration_ms = (time.time() - start_time) / 1000
```

Comment:
ANy specific reason why we need to have time in millisecond ? I think everywhere else in the codebase we use seconds ?

Reply:
This has been copied over from `demo/lightning/checkpoint/multinode/strategies.py`.